### PR TITLE
Pass tags from fastq to spring upon compression

### PR DESCRIPTION
**Pull request by islean**
## Description

Addresses issue #2131 wherein we want to reuse the fastq file tags when compressing to spring files.

### Changed

- Reuse the tags from one of the fastq files currently being compressed.

### How to prepare for test

- [x] Ssh to relevant server (depending on type of change)
- [x] Use stage: `us`
- [x] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_main -t cg -b pass-tags-fastq-spring -a
    ```

### How to test

- [ ] Do ...

### Expected test outcome

- [ ] Check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [x] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

**Review by diitaz93 — APPROVED** [review]
👍

**Review by henrikstranneheim — APPROVED** [review]
👍

**Review by seallard — APPROVED** [review]
Nice! Some lines could be extracted to separate functions to make it more readable

File: cg/meta/compress/compress.py [hunk 1]
```diff
@@ -223,8 +223,13 @@ def update_fastq_hk(
     ) -> None:
         """Update Housekeeper with compressed FASTQ files and SPRING metadata file."""
         version: Version = self.hk_api.last_version(sample_id)
-        spring_tags: List[str] = [sample_id, SequencingFileTag.SPRING]
-        spring_metadata_tags: List[str] = [sample_id, SequencingFileTag.SPRING_METADATA]
+        fastq_tags: List[str] = [
+            tag.name
+            for tag in self.hk_api.files(path=compression_obj.fastq_first.as_posix()).first().tags
+        ]
```
[the remark this answers is not in the document]
Comment (islean) [reply cg/meta/compress/compress.py]: Extracted!

**Comment by islean** [comment]
Test conducted by IO. Upon `cg compress clean fastq -c <case_id>`, the fastq files were deleted and the spring file was included with tags `[<flow_cell_id>, <SequencingFileTag.SPRING: 'spring'>]` and the spring metadata file was included with tags `[<flow_cell_id>, <SequencingFileTag.SPRING_METADATA: 'spring-metadata'>]`.

File: cg/meta/compress/compress.py [hunk 2]
```diff
@@ -259,6 +260,21 @@ def update_fastq_hk(
...
+    def get_spring_metadata_tags_from_fastq(self, fastq_file: File) -> List[str]:
+        spring_metadata_tags: List[str] = self.get_all_non_fastq_tags(fastq_file)
+        return spring_metadata_tags + [SequencingFileTag.SPRING_METADATA]
+
+    def get_spring_tags_from_fastq(self, fastq_file: File) -> List[str]:
+        spring_tags: List[str] = self.get_all_non_fastq_tags(fastq_file)
+        return spring_tags + [SequencingFileTag.SPRING]
+
+    @staticmethod
+    def get_all_non_fastq_tags(fastq_file: File) -> List[str]:
```
Comment (diitaz93) [inline cg/meta/compress/compress.py:272]: Beautiful 🌹

File: cg/meta/compress/compress.py [hunk 3]
```diff
@@ -259,6 +260,21 @@ def update_fastq_hk(
             hk_fastq_first=hk_fastq_first, hk_fastq_second=hk_fastq_second
         )
 
+    def get_spring_metadata_tags_from_fastq(self, fastq_file: File) -> List[str]:
+        spring_metadata_tags: List[str] = self.get_all_non_fastq_tags(fastq_file)
+        return spring_metadata_tags + [SequencingFileTag.SPRING_METADATA]
```
Comment (diitaz93) [inline cg/meta/compress/compress.py:265]:
```suggestion
        non_fastq_tags: List[str] = self.get_all_non_fastq_tags(fastq_file)
        return non_fastq_tags + [SequencingFileTag.SPRING_METADATA]
```

File: cg/meta/compress/compress.py [hunk 4]
```diff
@@ -259,6 +260,21 @@ def update_fastq_hk(
...
+
+    def get_spring_tags_from_fastq(self, fastq_file: File) -> List[str]:
+        spring_tags: List[str] = self.get_all_non_fastq_tags(fastq_file)
+        return spring_tags + [SequencingFileTag.SPRING]
```
Comment (diitaz93) [inline cg/meta/compress/compress.py:269]:
```suggestion
        non_fastq_tags: List[str] = self.get_all_non_fastq_tags(fastq_file)
        return non_fastq_tags + [SequencingFileTag.SPRING]
```

File: tests/meta/compress/test_clean_fastq.py [hunk 5]
```diff
@@ -95,6 +93,9 @@ def test_update_hk_fastq(
     )
     for spring_file in [hk_spring_files, hk_spring_metadata_files]:
         assert spring_file
+        for tag_name in [tag.name for tag in fastq[run]["hk_first"].tags]:
```
Comment (diitaz93) [inline tests/meta/compress/test_clean_fastq.py:96]: I don't understand this line. Shouldn't it be `spring_file.tags`?

**Review by diitaz93 — COMMENTED** [review]
Looks good!

File: tests/meta/compress/test_clean_fastq.py (hunk 5 above)
Comment (islean) [reply tests/meta/compress/test_clean_fastq.py:96]: @diitaz93 I wanted to make sure that all tags from the fastq file was passed on, so I loop through the fastq tags and ensure that the spring (metadata) file is tagged with them.

Comment (diitaz93) [reply tests/meta/compress/test_clean_fastq.py:96]: Is clear now 👍

**Comment by islean** [comment]
Deployed to stage and production on Hasta by IO, 2023-09-26:

```
Logging deploy ...
Getting deployer... done.
Getting last commit message and SHA... done.
Getting version of deploy scripts... /home/isak.ohlsson
done.
Log deploy... done.
cg, version 50.1.6
```

**Pull request merged**